src/lib/homestore/shard_manager.cpp

```diff
@@ -61,183 +62,119 @@ ShardManager::Result< ShardInfo > HSHomeObject::_create_shard(pg_id pg_owner, ui
     // 1. on the follower side
-    // 2. on the leader side but homeobject restarts and replay all commited log entries from the last commit id;
+    // 2. on the leader side but homeobject restarts and replay all commited log entries from the last checkpoint;
     if (user_ctx != nullptr) { promise = r_cast< folly::Promise< ShardManager::Result< ShardInfo > >* >(user_ctx); }
 
-    std::string shard_msg;
-    if (!precheck_and_decode_shard_msg(lsn, header, key, &shard_msg)) {
+    const ReplicationMessageHeader* msg_header = r_cast< const ReplicationMessageHeader* >(header.bytes);
+    if (msg_header->header_crc != msg_header->calculate_crc()) {
+        LOGWARN("replication message header is corrupted with crc error, lsn:{}", lsn);
         if (promise) { promise->setValue(folly::makeUnexpected(ShardError::INVALID_ARG)); }
         return;
     }
 
-    auto shard = deserialize_shard(shard_msg);
-    switch (shard.info.state) {
-    case ShardInfo::State::OPEN: {
-        std::scoped_lock lock_guard(_flying_shard_lock);
-        auto iter = _flying_shards.find(lsn);
-        if (iter == _flying_shards.end()) {
-            LOGWARN("can not find flying shards on lsn {}", lsn);
-            if (promise) { promise->setValue(folly::makeUnexpected(ShardError::UNKNOWN)); }
-            return;
-        }
-        shard.chunk_id = iter->second.chunk_id;
-        _flying_shards.erase(iter);
-        // serialize the finalized shard msg;
-        shard_msg = serialize_shard(shard);
-        // persist the serialize result to homestore MetaBlkService;
-        homestore::hs()->meta_service().add_sub_sb(HSHomeObject::s_shard_info_sub_type,
-                                                   r_cast< const uint8_t* >(shard_msg.c_str()), shard_msg.size(),
-                                                   shard.metablk_cookie);
-        // update in-memory shard map;
-        do_commit_new_shard(shard);
-        break;
-    }
-
-    case ShardInfo::State::SEALED: {
-        void* metablk_cookie = get_shard_metablk(shard.info.id);
-        RELEASE_ASSERT(metablk_cookie != nullptr, "seal shard when metablk is nullptr");
-        homestore::hs()->meta_service().update_sub_sb(r_cast< const uint8_t* >(shard_msg.c_str()), shard_msg.size(),
-                                                      metablk_cookie);
-        shard.metablk_cookie = metablk_cookie;
-        do_commit_seal_shard(shard);
-        break;
-    }
-    default: {
-        break;
-    }
-    }
-
-    if (promise) { promise->setValue(ShardManager::Result< ShardInfo >(shard.info)); }
+    // read value from PBA;
+    auto value_size = homestore::HomeStore::instance()->data_service().get_blk_size() * blkids.blk_count();
+    auto value_buf = iomanager.iobuf_alloc(512, value_size);
+    sisl::sg_list value;
+    value.size = value_size;
+    value.iovs.emplace_back(iovec{.iov_base = value_buf, .iov_len = value_size});
+
+    repl_dev.async_read(blkids, value, value_size)
```

Comment:
I believe we would be able to skip this `async_read` by providing `ShardInfo` into the context (a pinter to heap memory)?
And free the buffer after MetaSvc write the sb?

Reply:
yes, when switch to use ho_repl_ctx, no extra read is needed any more.